### ISSUE TITLE
Driver dispatch tests for MAC multipart

### ISSUE THREAD

**Issue by gilles-peskine-arm** [ARMmbed/mbedtls#5251]
Add driver dispatch tests for multipart MAC operations to `test_suite_psa_crypto_driver_wrappers`.

The new test functions should be similar to `cipher_{encrypt,decrypt}_multipart` and `hash_multipart_xxx`.

Take test data from `test_suite_psa_crypto.data` if available.

Ensure that all of the new tests are run by at least one `all.sh` component.

To be done in both development and 2.x.

Prerequisite: https://github.com/ARMmbed/mbedtls/pull/5089 and the 3.0 forward-port.

**Comment by mprse** [comment]
MAC (Message Authentication Code) can be can be calculated in two ways:
- single-part operation ( `psa_mac_compute()/psa_mac_verify()`), or
- multi-part operation ( `psa_mac_sign_setup()/psa_mac_verify_setup() -> psa_mac_update() called for each fragment ->  psa_mac_sign_finish()/psa_mac_verify_finish()`)

This task is about adding tests for multi-part operation, but I can see that there are some test that are testing also multi-part mac operations: `mac_sign`, `mac_verify`. Both functions are testing single and multi part operations. Multi-part is not tested well as `psa_mac_update()` is called only once. 

I understand that the plan is to create more detailed tests for MAC multipart operations:
- unit tests for each function (like `hash_multipart_xxx`): `mac_multipart_sign_setup`, `mac_multipart_verify_setup`, `mac_multipart_update`, `mac_multipart_sign_finish`, `mac_multipart_verify_finish`
- general test for multipart MAC calculations (like `cipher_{encrypt,decrypt}_multipart`): `mac_operation_multipart` 
and remove multi-part part from `mac_sign`, `mac_verify`?

**Comment by gilles-peskine-arm** [comment]
Thanks, I hadn't checked the test functions and so I hadn't noticed that they do test multipart operations.

So it's less urgent than I thought, but we still do want to have separate unit tests for one-shot and multipart functions eventually (since a driver might only provide one of the two sets), so we might as well do that now. Since the code is already there, we just need to split them.

We don't need individual unit tests for each function: e.g. a function like `update` can't be used on its own anyway, it can only be used after `setup` and you can't really know what it's done until a call to `finish`. For a multipart MAC, we only need to test:
* `mac_sign_setup`, `mac_update` (0/1/multiple times), `mac_sign_finish`
* `mac_sign_setup`, `mac_verify` (0/1/multiple times), `mac_verify_finish`
* any of the above interrupted at any point with a call to `mac_abort`

So far we aren't testing abort calls on most operation states. Adding tests for abort calls is outside the scope of this issue (but it is something we should add eventually, to all the multipart operations).

Testing multiple calls to `update` would be a good improvement. Also 0 calls (calculating the MAC of an empty message).

**Comment by gilles-peskine-arm** [comment]
Regarding abort calls in all operation states, I'm not sure we really need dedicated tests for the driver wrappers, as opposed to relying on the regular tests in builds with drivers. (And we're probably missing some regular tests of aborting halfway through an operation, but again that's out of scope here.)

**Comment by mprse** [comment]
Thanks for clarification. I understand that second bullet should be:
- `psa_mac_verify_setup`, `mac_update` (0/1/multiple times), `mac_verify_finish`
